chrome/content/zotero/xpcom/utilities.js

```diff
@@ -144,6 +144,238 @@ var CSL_TYPE_MAPPINGS = {
+		if (!identifier) {
+			// If no identifier is found, use other item fields to resolve it
+			let resolvedIdentifier = await Zotero.Utilities.resolveIdentifier(item);
+			// For now, resolve only DOI identifiers
+			if (resolvedIdentifier.DOI) {
+				identifier = resolvedIdentifier;
+			}
+		}
+		
+		let newItem = null;
+		
+		if (identifier) {
+			let item1 = await Zotero.Utilities.getItemFromAPI(identifier);
+			let item2 = null;
+			
+			if (identifier.DOI) {
+				item2 = await Zotero.Utilities.getItemFromURL('https://dx.doi.org/' + identifier.DOI);
+			}
+			
+			// If both items are translated
+			if (item1 && item2) {
+				// Combine item1 and item2
+				for (let key in item2) {
+					if (Zotero.Utilities.fieldIsValidForType(key, item1.itemType)) {
+						// Add all item2 fields that doesn't exist in item1
+						if (!item1[key] && item2[key]) {
+							item1[key] = item2[key];
+						}
+					}
+				}
+				newItem = item1;
+			}
+			// If only item1 is translated
+			else if (item1) {
+				newItem = item1;
+			}
+			// If only item2 is translated
+			else if (item2) {
+				newItem = item2;
+			}
+		}
+		
+		// Translate by URL if translation by identifier failed
+		if (!newItem && item.getField('url')) {
+			newItem = await Zotero.Utilities.getItemFromURL(item.getField('url'));
+		}
+		
+		let updated = false;
+		if (newItem) {
+			let newItemTypeID = Zotero.ItemTypes.getID(newItem.itemType);
+			
+			if (item.itemTypeID !== newItemTypeID) {
+				item.setType(newItemTypeID);
+				updated = true;
+			}
+			
+			let creators1 = item.toJSON().creators;
+			let creators2 = newItem.creators;
+			
+			let creatorsUpdated = false;
+			
+			// Compare creators count
+			if (creators1.length !== creators2.length) {
+				creatorsUpdated = true;
+			}
+			else {
+				// Compare each creator
+				for (let i = 0; i < creators1.length; i++) {
+					if (
+						creators1[i].creatorType !== creators2[i].creatorType ||
+						creators1[i].firstName !== creators2[i].firstName ||
+						creators1[i].lastName !== creators2[i].lastName
+					) {
+						creatorsUpdated = true;
+						break;
+					}
+				}
+			}
+			
+			if (creatorsUpdated) {
+				// Empty creators, because setCreators() doesn't do that bey default
+				item.setCreators([]);
+				item.setCreators(newItem.creators);
+				updated = true;
+			}
+			
+			// Get all field names for specific item type
+			let fields = Zotero.ItemFields.getItemTypeFields(Zotero.ItemTypes.getID(newItem.itemType));
+			fields = fields.map(x => Zotero.ItemFields.getName(x));
+			
+			// Transfer all fields from the new item to the existing one
+			for (let key of fields) {
+				let field1 = item.getField(key) || null;
+				let field2 = newItem[key] || null;
+				
+				if (field1 !== field2) {
+					// accessDate is always new, but it doesn't mean that metadata is changed
+					if (key !== 'accessDate') updated = true;
+					item.setField(key, field2);
+				}
+			}
```

Comment:
Is there a reason not to just use `fromJSON()` for all/most of the above?

Reply:
We are setting fields one by one because we don't want to overwrite the existing item with tags, notes and creation date. Therefore we only update fields that belong to specific item type. And also at the same time we are comparing fields to detect if the item was updated.